### PR TITLE
improved error handling and retry

### PR DESCRIPTION
**Pull request by xstelea**

### CODE REVIEW

File: src/application/test/open-api.test.ts [hunk 1]
```diff
@@ -11,19 +12,25 @@ const api = getAPI(openApiClient(new URL(BASE_URL)).call)
 const mock = new MockAdapter(axios)
 
 describe('handle error responses', () => {
+  beforeAll(() => {
```
Comment (AbstractFruitFactory) [inline src/application/test/open-api.test.ts:15]: don't think you need to do this before every test?

File: src/util/radixError.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+type RadixErrorInput = {
```
Comment (AbstractFruitFactory) [inline src/util/radixError.ts:1]: better do this inline if not used anywhere else

File: src/util/radixError.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  message: string
+  code?: number
+  details?: {
+    type: string
+  } & any
+  traceId?: string
+}
+
+export const radixError = (error: RadixErrorInput) => error
```
Comment (AbstractFruitFactory) [inline src/util/radixError.ts:10]: This is not a very useful function?

File: src/util/radixError.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+export type RadixError = ReturnType<typeof radixError>
```
Comment (AbstractFruitFactory) [inline src/util/radixError.ts:12]: this just exports the `RadixErrorInput` type right now - is the idea to build on this later somehow?

Comment (AbstractFruitFactory) [inline src/util/radixError.ts:12]:
```suggestion
export const radixError = (
  message: string,
  code?: number,
  details?: {
    type: string
  },
  traceId?: string
) => ({
  ...Error(message),
  code,
  details,
  traceId
})

export type RadixError = ReturnType<typeof radixError>
```

File: src/networking/open-api-client.ts [hunk 5]
```diff
@@ -43,19 +64,40 @@ export type ClientInterface = ReturnType<typeof AccountEndpointApiFactory> &
...
+  if (isNetworkError) {
+    return 'NetworkError'
+  }
+  switch (errorCode) {
+    case 'ECONNABORTED':
+      return 'RequestTimeoutError'
+
+    default:
+      return 'UnknownError'
+  }
```
Comment (AbstractFruitFactory) [inline src/networking/open-api-client.ts:78]:
```suggestion
if (message === 'Network Error') return 'NetworkError'
  return errorCode === 'ECONNABORTED' ? 'RequestTimeoutError' : 'UnknownError'
```

File: src/util/radixError.ts (hunk 4 above)
Comment (AbstractFruitFactory) [inline src/util/radixError.ts:12]:
```suggestion
enum ExceptionType {
  API = 'API',
  HARDWARE_WALLET = 'HardwareWallet',
  ERROR = 'Error'
}

type ApiExceptionParams = { code: number, type: string, traceId: string }
type HardwareWalletExceptionParams = {}

export const radixException = <
  T extends ExceptionType
>(
  message: string,
  type: T
) => <Params extends (
  T extends ExceptionType.API ? ApiExceptionParams :
  T extends ExceptionType.HARDWARE_WALLET ? HardwareWalletExceptionParams
  : never
)>(
  params: Params
) => ({
  ...Error(message),
  name: type,
  ...params,
})


// @ts-ignore
const apiWrapper = radixException<ExceptionType.API>()
type APIException = ReturnType<typeof apiWrapper>

// @ts-ignore
const hwWrapper = radixException<ExceptionType.HARDWARE_WALLET>()
type HWException = ReturnType<typeof hwWrapper>

export type RadixException = APIException | HWException


```

**Comment by AbstractFruitFactory** [comment]
We need to be able to easily define errors of different types, e.g API errors and hardware wallet errors. Also, we lose the stack trace if we don't use the native `Error` object. See code suggestion

File: src/application/test/open-api.test.ts (hunk 1 above)
Comment (xstelea) [reply src/application/test/open-api.test.ts:15]: No, it's working fine. We use the same setup in integration tests

**Pull request merged**